src/main/scala/org/neo4j/spark/service/SchemaService.scala

```diff
@@ -331,18 +331,102 @@ class SchemaService(private val options: Neo4jOptions, private val driverCache:
     }
   }
 
-  def isReadQuery(query: String): Boolean = {
+  def isQueryType(query: String, expectedQueryTypes: org.neo4j.driver.summary.QueryType*): Boolean = try {
     val queryType = session.run(s"EXPLAIN $query").consume().queryType()
-    queryType == org.neo4j.driver.summary.QueryType.READ_ONLY || queryType == org.neo4j.driver.summary.QueryType.SCHEMA_WRITE
+    expectedQueryTypes.contains(queryType)
+  } catch {
+    case e: Throwable => {
+      log.error("Query not compiled because of the following exception:", e)
+      false
+    }
   }
 
   def isValidQueryCount(query: String): Boolean = {
-    val resultSummary = session.run(s"EXPLAIN $query").consume()
-    val queryType = resultSummary.queryType()
-    val plan = resultSummary.plan()
-    val isReadOnly = queryType == org.neo4j.driver.summary.QueryType.READ_ONLY || queryType == org.neo4j.driver.summary.QueryType.SCHEMA_WRITE
-    val hasCountIdentifier = plan.identifiers().asScala.toSet == Set("count")
-    isReadOnly && hasCountIdentifier
+    try {
+      val resultSummary = session.run(s"EXPLAIN $query").consume()
+      val queryType = resultSummary.queryType()
+      val plan = resultSummary.plan()
+      val isReadOnly = queryType == org.neo4j.driver.summary.QueryType.READ_ONLY || queryType == org.neo4j.driver.summary.QueryType.SCHEMA_WRITE
+      val hasCountIdentifier = plan.identifiers().asScala.toSet == Set("count")
+      isReadOnly && hasCountIdentifier
+    } catch {
+      case e: Throwable => {
+        log.error("Query not compiled because of the following exception:", e)
+        false
+      }
+    }
+  }
+
+  private def createIndexOrConstraint(action: OptimizationType.Value, label: String, props: Seq[String]) = {
+    if (action == OptimizationType.NONE) {
+      log.info("No optimization type provided")
+      Unit
+    } else {
+      val quotedLabel = label.quote()
+      val quotedProps = props.map(prop => s"${Neo4jUtil.NODE_ALIAS}.${prop.quote()}").mkString(", ")
+      val actionName = s"spark_${action}_$label".quote()
+      val queryPrefix = s"CREATE ${action.toString} $actionName"
+      val querySuffix = action match {
+        case OptimizationType.INDEX => s"FOR (${Neo4jUtil.NODE_ALIAS}:$quotedLabel) ON ($quotedProps)"
+        case OptimizationType.CONSTRAINT => {
+          val assertType = if (props.size > 1) "NODE KEY" else "UNIQUE"
+          s"ON (${Neo4jUtil.NODE_ALIAS}:$quotedLabel) ASSERT ($quotedProps) IS $assertType"
+        }
+      }
+      val status = try {
+        session.run(s"$queryPrefix $querySuffix")
+        "CREATED"
+      } catch {
+        case _: Throwable => "KEPT"
```

Comment:
not sure how this exception works but in 4.0 indexes fail to create if you do it twice, is this trying to catch the case where the index already existed?  In 3.5, index creation is idempotent

Reply:
I can change it by adding an existence check on top, it will work on `3.5` and `4.x`